homeassistant/components/sensor/monzo.py

```diff
@@ -0,0 +1,329 @@
+"""Monzo Sensor.
+
+Used to create a sensor in Home Assistant that will return
+the current balance of a specified monzo account.
+"""
+import logging
+import json
+import time
+import urllib.parse
+import random
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.core import callback
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.const import (
+    CONF_NAME)
+from requests import post
+
+REQUIREMENTS = ['monzo==0.5.3']
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = 'mdi:credit-card'
+
+
+DEFAULT_CACHE_PATH = '.monzo-token-cache'
+AUTH_CALLBACK_PATH = '/api/monzo'
+AUTH_CALLBACK_NAME = 'api:monzo'
+DEFAULT_NAME = 'Monzo Balance'
+DOMAIN = 'monzo'
+CONF_CLIENT_ID = 'client_id'
+CONF_CLIENT_SECRET = 'client_secret'
+CONF_CACHE_PATH = 'cache_path'
+CONF_CURRENT_ACCOUNT = 'current_account'
+
+CONFIGURATOR_LINK_NAME = 'Link Monzo account'
+CONFIGURATOR_SUBMIT_CAPTION = 'I authorized successfully'
+CONFIGURATOR_DESCRIPTION = 'To link your Monzo account, ' \
+                           'click the link, login, and authorize:'
+
+
+def is_token_expired(token_info):
+    """Return true if the token has expired."""
+    now = int(time.time())
+    return token_info['expires_at'] - now < 60
+
+
+def request_configuration(hass, config, add_devices, request_url):
+    """Request Monzo authorization."""
+    configurator = hass.components.configurator
+    hass.data[DOMAIN] = configurator.request_config(
+        DEFAULT_NAME, lambda _: None,
+        link_name=CONFIGURATOR_LINK_NAME,
+        link_url=request_url,
+        description=CONFIGURATOR_DESCRIPTION,
+        submit_caption=CONFIGURATOR_SUBMIT_CAPTION)
+
+
+def setup_platform(hass, config, add_devices, device_discovery=None):
+    """Setup the monzo platform."""
+    client_id = config.get(CONF_CLIENT_ID)
+    client_secret = config.get(CONF_CLIENT_SECRET)
+    callback_url = '{}{}'.format(hass.config.api.base_url, AUTH_CALLBACK_PATH)
+    cache = config.get(CONF_CACHE_PATH, hass.config.path(DEFAULT_CACHE_PATH))
+    current_account = config.get(CONF_CURRENT_ACCOUNT, False)
+    oauth = AuthClient(client_id,
+                       client_secret,
+                       callback_url,
+                       cache_path=cache)
+
+    token_info = oauth.get_cached_token()
+    if not token_info:
+        hass.http.register_view(MonzoAuthCallbackView(
+            config, add_devices, oauth))
+        request_configuration(hass,
+                              config,
+                              add_devices,
+                              oauth.get_authorize_url())
+        return
+    if hass.data.get(DOMAIN):
+        configurator = hass.components.configurator
+        configurator.request_done(hass.data.get(DOMAIN))
+        del hass.data[DOMAIN]
+    sensor = MonzoSensor(oauth,
+                         current_account,
+                         config.get(CONF_NAME, DEFAULT_NAME))
+    add_devices([sensor])
+
+
+class AuthClient():
```

Comment:
This should be handled by `monzo`. The module should be extend if there is no support for this kind of functionality.